src/coreclr/jit/morphblock.cpp

```diff
@@ -0,0 +1,1551 @@
+        m_dstLclOffset = m_dstLclNode->GetLclOffs();
+
+#if LOCAL_ASSERTION_PROP
+
+        // Kill everything about m_dstLclNum (and its field locals)
+        if (m_comp->optLocalAssertionProp && (m_comp->optAssertionCount > 0))
+        {
+            m_comp->fgKillDependentAssertions(m_dstLclNum DEBUGARG(m_asg));
+        }
+
+#endif // LOCAL_ASSERTION_PROP
+    }
+
+#if defined(DEBUG)
+    if (m_comp->verbose)
+    {
+        printf("PrepareDst for [%06u] ", m_comp->dspTreeID(origDst));
+        if (m_dstLclNode != nullptr)
+        {
+            printf("have found a local var V%02u.\n", m_dstLclNum);
+        }
+        else
+        {
+            printf("have not found a local var.\n");
+        }
+    }
+#endif // DEBUG
+}
+
+//------------------------------------------------------------------------
+// PrepareSrc: Transform the asg src to an appropriate form and initialize member fields
+//    with information about it.
+//
+void MorphInitBlockHelper::PrepareSrc()
+{
+    m_src = m_asg->gtGetOp2();
+}
+
+//------------------------------------------------------------------------
+// TrySpecialCases: check special cases that require special transformations.
+//    We don't have any for for init block.
+//
+void MorphInitBlockHelper::TrySpecialCases()
+{
+    return;
+}
+
+//------------------------------------------------------------------------
+// MorphStructCases: transforms the asg as field by field init or keeps it as a block init
+//    but sets appropriate flags for the involved lclVars.
+//
+// Assumptions:
+//    we have already checked that it is not a special case and can't be transformed as OneAsgBlock.
+//
+void MorphInitBlockHelper::MorphStructCases()
+{
+    GenTree* initVal = m_src->OperIsInitVal() ? m_src->gtGetOp1() : m_src;
+
+    if (m_dstLclNum != BAD_VAR_NUM)
+    {
+        // If we have already determined that a promoted TYP_STRUCT lclVar will not be enregistered,
+        // we are better off doing a block init.
+        bool tryFieldByField = false;
+        if (m_dstVarDsc->lvPromoted)
+        {
+            assert(m_dstLclNode->OperIs(GT_LCL_VAR)); // Don't expect a promoted LCL_VAR with a field reference.
+            if (m_dstVarDsc->lvDoNotEnregister)
+            {
+                tryFieldByField = false;
+            }
+            else
+            {
+                tryFieldByField = true;
+            }
+        }
+
+        if (tryFieldByField)
+        {
+            GenTreeLclVar* dstLclVar = m_dstLclNode->AsLclVar();
+            GenTree*       newTree   = m_comp->fgMorphPromoteLocalInitBlock(dstLclVar, initVal, m_blockSize);
+
+            if (newTree != nullptr)
+            {
+                m_transformationDecision = BlockTransformation::FieldByField;
+                m_result                 = newTree;
+            }
+        }
+
+        if (m_transformationDecision != BlockTransformation::FieldByField)
+        {
+            if (m_dst != m_dstLclNode)
```

Comment:
this is an important change, in the past for `ASG(struct<8> notPromotedStruct, 0)` if we did not transform it as OneAsg (for example, the struct has holes and a custom layout) we would set `lvaSetVarDoNotEnregister`. Now we don't, LSRA won't try to enregister it because it can't enregister structs, but if we enable `JitEnregStructLocals=1` we will start to see the diffs that we wanted. https://github.com/dotnet/runtime/pull/52802 was an important step to make this pessimization deletion possible.